Unable to retrieve the value of crc32 command in uboot with exec0.

Hi!

This assertion is failing because `exec0` does not return the complete result, even though stdout does!
Could you please advise me on how to ensure that the entire result is returned?

```          
            .
            .
            pattern = re.compile(r'=(\w+)')
            fileaddr =  pattern.findall(ub.env("fileaddr"))
            filesize =  pattern.findall(ub.env("filesize"))
            output = ub.exec0("crc32", f"{fileaddr[0]}" ,f"{filesize[0]}")
>           assert expected_crc in output
E           AssertionError: assert 'b2aff8d0' in '\ncrc32 for 80280000 ... 8028004b ='
```

```
│   ├─[device-u-boot] crc32 80280000 4c
│   │    ## 
│   │    ## crc32 for 80280000 ... 8028004b ==> b2aff8d0
```

--- a/selftest/testmachines.py
+++ b/selftest/testmachines.py
@@ -272,6 +272,11 @@ function setenv() {
     shift
     eval "$var=\\"$*\\""
 }
+function crc32() {
+    printf "crc32 for %s ... %s ==> " "$1" "$1"
+    sleep 0.2
+    echo "deadb33f"
+}
 function boot() {
     echo "Pretending to boot Linux..."
     echo ""

--- a/selftest/tests/test_board.py
+++ b/selftest/tests/test_board.py
@@ -59,6 +59,14 @@ def test_uboot_simple_control(tbot_context: tbot.Context) -> None:
         assert out == "FOO"
 
 
+def test_uboot_crc32_workaround(tbot_context: tbot.Context) -> None:
+    with tbot_context.request(testmachines.MockhwBoardUBoot) as ub:
+        if ub.prompt != "=> ":
+            pytest.skip("Wrong U-Boot prompt for crc32 workaround test.")
+        ub.exec0("crc32", "0x10000008", "0x42")
+        assert ub.exec0("echo", "Hello World") == "Hello World\n"
+
+
 def test_linux_boot(tbot_context: tbot.Context) -> None:
     with tbot_context.request(testmachines.MockhwBoardLinux) as lnx:
         out = lnx.exec0("echo", "Hello World")

--- a/tbot/machine/board/uboot.py
+++ b/tbot/machine/board/uboot.py
@@ -269,10 +269,29 @@ class UBootShell(shell.Shell, UbootStartup):
         """
         cmd = self.escape(*args)
 
+        # There is an ugly ugly problem with no great solution: The `crc32`
+        # command in U-Boot prints the string `=> ` as part of its output.
+        # This is a commonly used prompt string which means that tbot gets
+        # completely thrown off by this situation.  As a stopgap solution,
+        # let's special case this here to make people's lives a bit easier...
+        if args[0] == "crc32" and self.ch.prompt in ("=> ", b"=> "):
+            tbot.log.warning(
+                "Applying workaround for `crc32` command in U-Boot!\n"
+                + "  See https://github.com/rahix/tbot/issues/111 for details."
+            )
+            override_prompt = "\n=> "
+        else:
+            override_prompt = None
+
         with tbot.log_event.command(self.name, cmd) as ev:
             self.ch.sendline(cmd, read_back=True)
-            with self.ch.with_stream(ev, show_prompt=False):
-                out = self.ch.read_until_prompt()
+            with self.ch.with_prompt(override_prompt):
+                with self.ch.with_stream(ev, show_prompt=False):
+                    out = self.ch.read_until_prompt(prompt=override_prompt)
+                    if override_prompt == "\n=> ":
+                        # The overridden prompt ate the trailing '\n'
+                        ev.write("\n")
+                        out += "\n"
             ev.data["stdout"] = out
 
             self.ch.sendline("echo $?", read_back=True)

--- a/tbot/machine/channel/channel.py
+++ b/tbot/machine/channel/channel.py
@@ -841,7 +841,7 @@ class Channel(typing.ContextManager):
     # prompt handling {{{
     @contextlib.contextmanager
     def with_prompt(
-        self, prompt_in: ConvenientSearchString
+        self, prompt_in: typing.Optional[ConvenientSearchString]
     ) -> "typing.Iterator[Channel]":
         """
         Set the prompt for this channel during a context.
@@ -859,22 +859,27 @@ class Channel(typing.ContextManager):
         :param ConvenientSearchString prompt: The new prompt pattern/string.
             See :ref:`channel_search_string` for more info.
         """
-        prompt = _convert_search_string(prompt_in)
-
-        # If the prompt is a pattern, we need to recompile it with an additional $ in the
-        # end to ensure that it only matches the end of the stream
-        if isinstance(prompt, BoundedPattern):
-            new_pattern = re.compile(
-                prompt.pattern.pattern + b"$", prompt.pattern.flags
-            )
-            prompt = BoundedPattern(new_pattern)
-
-        previous = self.prompt
-        self.prompt = prompt
-        try:
+        if prompt_in is None:
+            # Ignore when no prompt was given. This makes conditional prompt
+            # overrides easier to implement.
             yield self
-        finally:
-            self.prompt = previous
+        else:
+            prompt = _convert_search_string(prompt_in)
+
+            # If the prompt is a pattern, we need to recompile it with an additional $ in the
+            # end to ensure that it only matches the end of the stream
+            if isinstance(prompt, BoundedPattern):
+                new_pattern = re.compile(
+                    prompt.pattern.pattern + b"$", prompt.pattern.flags
+                )
+                prompt = BoundedPattern(new_pattern)
+
+            previous = self.prompt
+            self.prompt = prompt
+            try:
+                yield self
+            finally:
+                self.prompt = previous
 
     def read_until_prompt(
         self,

